File: nfc_proof_of_concept/packages/nfc_host_card_emulation/lib/app_layer/file_access/serializers/apdu_response_serializer.dart
import 'dart:typed_data';

import '../../field.dart';
import '../fields/response_fields.dart';

class ApduResponse extends ApduSerializer {
  final ApduData? data;
  final ApduStatusWord statusWord;

  ApduResponse._internal({
    this.data,
    required this.statusWord,
    required String name,
  }) : super(name: name);

  factory ApduResponse.fromBytes(Uint8List rawResponse) {
    if (rawResponse.length < 2) {
      throw ArgumentError('Invalid APDU response: must be at least 2 bytes for the status word. Got ${rawResponse.length} bytes.');
    }

    final dataLength = rawResponse.length - 2;
    final responseData = (dataLength > 0)
        ? ApduData(rawResponse.sublist(0, dataLength), name: "Response Data (Parsed)")
        : null;

    final sw1 = rawResponse[dataLength];
    final sw2 = rawResponse[dataLength + 1];
    final statusWord = ApduStatusWord.fromBytes(sw1, sw2);
    
    return ApduResponse._internal(
      data: responseData,
      statusWord: statusWord,
      name: "Parsed Response",
    );
  }

  factory ApduResponse.success({Uint8List? data}) {
    return ApduResponse._internal(
      data: (data != null && data.isNotEmpty)
          ? ApduData(data, name: "Response Data")
          : null,
      statusWord: ApduStatusWord.ok,
      name: "Success Response",
    );
  }

  /// Creates an error response with the specified status word.
  factory ApduResponse.error(ApduStatusWord errorStatus) {
    if (errorStatus == ApduStatusWord.ok) {
      throw ArgumentError('Cannot create an error response with SW=9000. Use success() factory.');
    }
    return ApduResponse._internal(
      statusWord: errorStatus,
      name: "Error Response",
    );
  }

  @override
  void setFields() {
    fields = [
      data,
      statusWord,
    ];
  }
}

File: nfc_proof_of_concept/packages/flutter_hce/android/src/main/kotlin/com/viridian/flutter_hce/app_layer/file_access/serializers/ApduResponseSerializer.kt
package com.viridian.flutter_hce.app_layer.file_access.serializers

import com.viridian.flutter_hce.app_layer.ApduData
import com.viridian.flutter_hce.app_layer.ApduSerializer
import com.viridian.flutter_hce.app_layer.Bytes
import com.viridian.flutter_hce.app_layer.file_access.fields.ApduStatusWord

class ApduResponse private constructor(
    private val data: ApduData?,
    private val statusWord: ApduStatusWord,
    name: String
) : ApduSerializer(name) {

    companion object {
        fun fromBytes(rawResponse: Bytes): ApduResponse {
            require(rawResponse.size >= 2) { 
                "Invalid APDU response: must be at least 2 bytes for the status word. Got ${rawResponse.size} bytes." 
            }

            val dataLength = rawResponse.size - 2
            val responseData = if (dataLength > 0) {
                ApduData(rawResponse.sliceArray(0 until dataLength), "Response Data (Parsed)")
            } else null

            val sw1 = rawResponse[dataLength].toInt() and 0xFF
            val sw2 = rawResponse[dataLength + 1].toInt() and 0xFF
            val statusWord = ApduStatusWord.fromBytes(sw1, sw2)

            return ApduResponse(responseData, statusWord, "Parsed Response")
        }

        fun success(data: Bytes? = null): ApduResponse {
            val responseData = if (data != null && data.isNotEmpty()) {
                ApduData(data, "Response Data")
            } else null

            return ApduResponse(responseData, ApduStatusWord.ok, "Success Response")
        }

        fun error(errorStatus: ApduStatusWord): ApduResponse {
            require(errorStatus != ApduStatusWord.ok) { 
                "Cannot create an error response with SW=9000. Use success() factory." 
            }
            return ApduResponse(null, errorStatus, "Error Response")
        }
    }

    override fun setFields() {
        fields.clear()
        fields.addAll(listOfNotNull(data, statusWord))
    }
}
